src/map.cpp

```diff
@@ -212,12 +212,31 @@ VehicleList map::get_vehicles()
                          tripoint( SEEX * my_MAPSIZE, SEEY * my_MAPSIZE, OVERMAP_HEIGHT ) );
 }
 
-void map::reset_vehicle_cache( const int zlev )
+void map::clear_all_vehicle_caches( const int zlev )
 {
-    clear_vehicle_cache( zlev );
-    // Cache all vehicles
-    auto &ch = get_cache( zlev );
-    ch.veh_in_active_range = false;
+    const int zmin = zlevels ? -OVERMAP_DEPTH : zlev;
+    const int zmax = zlevels ? OVERMAP_HEIGHT : zlev;
+
+    for( int gridz = zmin; gridz <= zmax; gridz++ ) {
+        level_cache &ch = get_cache( gridz );
+        while( !ch.veh_cached_parts.empty() ) {
+            const auto part = ch.veh_cached_parts.begin();
+            const auto &p = part->first;
+            if( inbounds( p ) ) {
+                ch.veh_exists_at[p.x][p.y] = false;
+            }
+            ch.veh_cached_parts.erase( part );
+        }
+        ch.veh_in_active_range = false;
+    }
+}
+
+void map::rebuild_vehicle_cache( const int zlev )
```

Comment:
Since now `rebuild_vehicle_cache` doesn't clear the cache before building, perhaps rename the method to `build_vehicle_cache` to emphasize that?